hera_cal/lstbin_simple.py

```diff
@@ -0,0 +1,1677 @@
+    data
+        The visibility data. Must be shape (ntimes, nbls, nfreqs, npols)
+    data_lsts
+        The LSTs corresponding to each of the time stamps in the data. Must have
+        length ``data.shape[0]``
+    baselines
+        The list 2-tuples of baselines in the data array.
+    lst_bin_edges
+        A sequence of floats specifying the *edges* of the LST bins to use.
+    freq_array
+        An array of frequencies in the data, in Hz.
+    flags
+        An array of boolean flags, indicating bins NOT to use. Same shape as data.
+    nsamples
+        An array of sample counts, same shape as data.
+    rephase
+        Whether to apply re-phasing to the data, to bring it to a common LST grid.
+    antpos
+        3D Antenna positions for each antenna in the data.
+    lat
+        The latitude (in degrees) of the telescope.
+
+    Returns
+    -------
+    data
+        A nlst-length list of arrays, each of shape 
+        ``(ntimes_in_lst, nbls, nfreq, npol)``, where LST bins without data simply have
+        a first-axis of size zero.
+    flags
+        Same as ``data``, but boolean flags.
+    nsamples
+        Same as ``data``, but sample counts.
+
+    See Also
+    --------
+    :func:`reduce_lst_bins`
+        Function that takes outputs from this function and computes reduced values (e.g.
+        mean, std) from them.
+    """
+    npols = data.shape[-1]
+    required_shape = (len(data_lsts), len(baselines), len(freq_array), npols)
+    
+    if npols > 4:
+        raise ValueError(f"data has more than 4 pols! Got {npols} (last axis of data)")
+
+    if data.shape != required_shape:
+        raise ValueError(
+            f"data should have shape {required_shape} but got {data.shape}"
+        )
+
+    if flags is None:
+        flags = np.zeros(data.shape, dtype=bool)
+
+    if flags.shape != data.shape:
+        raise ValueError(
+            f"flags should have shape {data.shape} but got {flags.shape}"
+        )
+
+    if nsamples is None:
+        nsamples = np.ones(data.shape, dtype=float)
+    
+    if nsamples.shape != data.shape:
+        raise ValueError(
+            f"nsamples should have shape {data.shape} but got {nsamples.shape}"
+        )
+
+    if len(lst_bin_edges) < 2:
+        raise ValueError("lst_bin_edges must have at least 2 elements")
+
+    # Ensure the lst bin edges start within (0, 2pi)
+    adjust_lst_bin_edges(lst_bin_edges)
+
+    if not np.all(np.diff(lst_bin_edges) > 0):
+        raise ValueError(
+            "lst_bin_edges must be monotonically increasing."
+        )
+
+    # Now ensure that all the observed LSTs are wrapped so they start above the first bin edges
+    grid_indices, data_lsts, lst_mask = get_lst_bins(data_lsts, lst_bin_edges)
+    lst_bin_centres = (lst_bin_edges[1:] + lst_bin_edges[:-1])/2
+
+    # TODO: check whether this creates a data copy. Don't want the extra RAM...
+    data = data[lst_mask]  # actually good if this is copied, because we do LST rephase in-place
```

Comment:
can we rectify these comments, and if you do want this to be a copy let's make this explicit via deepcopy

Reply:
I determined what's actually going on, and wrote more extensive comments in-line. The `lst_mask` indexing does do a copy, so we don't want to `deepcopy` because that'd be yet another copy.